modules/server/src/main/scala/scaladex/server/service/UserSessionSynchronizer.scala

```diff
@@ -0,0 +1,65 @@
+package scaladex.server.service
+
+import java.util.UUID
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.concurrent.duration.DurationInt
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.model.StatusCodes
+import cats.implicits.toTraverseOps
+import scaladex.core.model.GithubResponse
+import scaladex.core.model.UserInfo
+import scaladex.core.model.UserState
+import scaladex.core.service.GithubService
+import scaladex.core.service.SchedulerDatabase
+import scaladex.infra.GithubClient
+
+class UserSessionSynchronizer(database: SchedulerDatabase)(implicit ec: ExecutionContext, ac: ActorSystem)
+    extends Scheduler("user-session-synchronizer", 1.hour) {
+
+  type Session = (UUID, UserState)
+
+  override def run(): Future[Unit] =
+    for {
+      sessions <- database.getAllSessions()
+      updatedEitherUserSessions <- sessions.traverse { case (session, client) => updateUserSession(session, client) }
+      expiredSessionIds = updatedEitherUserSessions.collect { case Left(expiredSessionId) => expiredSessionId }
+      updatedUserSessions = updatedEitherUserSessions.collect { case Right(sessions) => sessions }
+      _ <- updatedUserSessions.traverse { case (userId, userState) => database.insertSession(userId, userState) }
+      _ <- expiredSessionIds.traverse(database.deleteSession)
+    } yield ()
+
+  private def updateUserSession(session: Session): Future[Either[UUID, Session]] =
+    session match {
+      case (userId, staleUserState) =>
+        val service = new GithubClient(staleUserState.info.token)
+        getUserInfo(userId, staleUserState.info, service)
+          .map(_.map { updatedUserInfo =>
+            if (updatedUserInfo != staleUserState.info)
+              (userId, staleUserState.copy(info = updatedUserInfo))
+            else session
+          })
```

Comment:
What if you match on `GithubResponse` here and make the call to the database directly:

```suggestion
          service.getUserState().flatMap  {
            case GithubResponse.Ok(state) => database.insertSession(userId, state)
            case GithubResponse.MovedPermanently(state) => database.insertSession(userId, state)
            case GithubResponse.Failed(code, errorMessage) =>
              if (code == StatusCode.Unauthorized.intValue) {
                logger.info(s"Token for user with id: '$userId' is likely expired, with error: $errorMessage")
                database.deleteSession(userId)
              } else {
                logger.warn(s"Failed to get user state from Github because of $code: $errorMessage")
                Future.successful(()) // we cannot update the user state
              }
          }
```